core/services/keeper/upkeep_executer.go

```diff
@@ -0,0 +1,204 @@
+}
+
+func (executor *UpkeepExecutor) Close() error {
+	if !executor.OkayToStop() {
+		return errors.New("UpkeepExecutor is already stopped")
+	}
+	close(executor.chDone)
+	executor.doneWG.Wait()
+	return nil
+}
+
+func (executor *UpkeepExecutor) Connect(head *models.Head) error {
+	return nil
+}
+
+func (executor *UpkeepExecutor) Disconnect() {}
+
+func (executor *UpkeepExecutor) OnNewLongestChain(ctx context.Context, head models.Head) {
+	executor.blockHeight.Store(head.Number)
+	// avoid blocking if signal already in buffer
+	select {
+	case executor.chSignalRun <- struct{}{}:
+	default:
+	}
+}
+
+func (executor *UpkeepExecutor) run() {
+	executor.doneWG.Add(1)
+	for {
+		select {
+		case <-executor.chDone:
+			executor.doneWG.Done()
+			return
+		case <-executor.chSignalRun:
+			executor.processActiveUpkeeps()
+		}
+	}
+}
+
+func (executor *UpkeepExecutor) processActiveUpkeeps() {
+	// Keepers could miss their turn in the turn taking algo if they are too overloaded
+	// with work because processActiveUpkeeps() blocks
+	blockHeight := executor.blockHeight.Load()
+	logger.Debug("received new block, running checkUpkeep for keeper registrations", "blockheight", blockHeight)
+
+	ctx, _ := postgres.DefaultQueryCtx()
+	activeUpkeeps, err := executor.orm.EligibleUpkeeps(ctx, blockHeight)
+	if err != nil {
+		logger.Errorf("unable to load active registrations: %v", err)
+		return
+	}
+
+	wg := sync.WaitGroup{}
+	wg.Add(len(activeUpkeeps))
+	// TODO - RYAN - comment
+	done := func() { <-executor.executionQueue; wg.Done() }
+	for _, reg := range activeUpkeeps {
+		executor.executionQueue <- struct{}{}
+		go executor.execute(reg, done)
+	}
+
+	wg.Wait()
+}
+
+// execute will call checkForUpkeep and, if it succeeds, triger a job on the CL node
+// DEV: must perform contract call "manually" because abigen wrapper can only send tx
+func (executor *UpkeepExecutor) execute(upkeep UpkeepRegistration, done func()) {
+	defer done()
+
+	msg, err := constructCheckUpkeepCallMsg(upkeep)
+	if err != nil {
+		logger.Error(err)
+		return
+	}
+
+	logger.Debugf("Checking upkeep on registry: %s, upkeepID %d", upkeep.Registry.ContractAddress.Hex(), upkeep.UpkeepID)
+
+	checkUpkeepResult, err := executor.ethClient.CallContract(context.Background(), msg, nil)
+	if err != nil {
+		logger.Debugf("checkUpkeep failed on registry: %s, upkeepID %d", upkeep.Registry.ContractAddress.Hex(), upkeep.UpkeepID)
+		return
+	}
+
+	performTxData, err := constructPerformUpkeepTxData(checkUpkeepResult, upkeep.UpkeepID)
+	if err != nil {
+		logger.Error(err)
+		return
+	}
+
+	logger.Debugf("Performing upkeep on registry: %s, upkeepID %d", upkeep.Registry.ContractAddress.Hex(), upkeep.UpkeepID)
+
+	ctx, _ := postgres.DefaultQueryCtx()
+	err = executor.orm.CreateEthTransactionForUpkeep(ctx, upkeep, performTxData)
+	if err != nil {
+		logger.Error(err)
+	}
+}
+
+func constructCheckUpkeepCallMsg(upkeep UpkeepRegistration) (ethereum.CallMsg, error) {
+	checkPayload, err := RegistryABI.Pack(
```

Comment:
Is there no way to use the abigen wrapper to do this? If not, maybe we can add another enhancement to the abigen enhancer?

Reply:
The wrapper can pack and unpack the args for us, but it can't make an eth call because the solidity function is not a `view` function. I think this a rare enough use case that an abigen enhancement is probably not necessary.